ddl/index.go

```diff
@@ -700,6 +685,432 @@ func (w *worker) doBackfillIndexTaskInTxn(t table.Table, txn kv.Transaction, col
 	return taskRet
 }
 
+type addIndexWorker struct {
+	id            int
+	d             *ddl
+	batchCnt      int
+	sessCtx       sessionctx.Context
+	taskCh        chan *reorgIndexTask
+	resultCh      chan *addIndexResult
+	notRunnableCh chan struct{}
+	index         table.Index
+	table         table.Table
+	colFieldMap   map[int64]*types.FieldType
+	closed        bool
+
+	defaultVals []types.Datum         // It's used to reduce the number of new slice.
+	idxRecords  []*indexRecord        // It's used to reduce the number of new slice.
+	rowMap      map[int64]types.Datum // It's the index column values map. It is used to reduce the number of making map.
+}
+
+type reorgIndexTask struct {
+	startHandle int64
+	endHandle   int64
+}
+
+type addIndexResult struct {
+	addedCount int
+	scanCount  int
+	nextHandle int64
+	err        error
+}
+
+func newAddIndexWoker(sessCtx sessionctx.Context, d *ddl, id int, index table.Index, t table.Table, colFieldMap map[int64]*types.FieldType) *addIndexWorker {
+	return &addIndexWorker{
+		id:            id,
+		d:             d,
+		batchCnt:      defaultTaskHandleCnt,
+		sessCtx:       sessCtx,
+		taskCh:        make(chan *reorgIndexTask),
+		resultCh:      make(chan *addIndexResult),
+		notRunnableCh: make(chan struct{}, 1),
+		index:         index,
+		table:         t,
+		colFieldMap:   colFieldMap,
+
+		defaultVals: make([]types.Datum, len(t.Cols())),
+		rowMap:      make(map[int64]types.Datum, len(colFieldMap)),
+	}
+}
+
+func (w *addIndexWorker) isWorkerRunnable() bool {
+	select {
+	case <-w.notRunnableCh:
+		return false
+	default:
+	}
+	return true
+}
+
+func (w *addIndexWorker) notifyNotRunnable() {
+	select {
+	case w.notRunnableCh <- struct{}{}:
+	default:
+	}
+}
+
+func (w *addIndexWorker) close() {
+	if !w.closed {
+		w.closed = true
+		close(w.taskCh)
+	}
+}
+
+// getIndexRecord gets index columns values from raw binary value row.
+func (w *addIndexWorker) getIndexRecord(t table.Table, colMap map[int64]*types.FieldType, rawRecord []byte, idxRecord *indexRecord) error {
```

Comment:
this function is copied from `worker. getIndexRecord`, I will remove the worker. getIndexRecord later.

Reply:
`t` is stored in the worker and `colMap` can be stored in the worker.
And I think accept `handle` and `recordKey`, return `*indexRecord` is better.